### ISSUE TITLE
Transfer Results on CIFAR-10

### ISSUE THREAD

**Issue by YUE-FAN** [SHI-Labs/Semi-Supervised-Transfer-Learning#4]
Hi,

Thanks for the great work. I have a quick question regarding the results in table 7.

Directly transferring your training protocol from CUB 200, I managed to reproduce your MixMatch results on CIFAR-10. But my attempt to reproduce the 9.05% test accuracy on CIFAR-10 with 40 labels using FixMatch was not successful. I think I must have missed some important hyper-parameters. Would you mind provide more details on that experiment?

**Comment by Walleclipse** [comment]
Hi,
Hyper-parameters of the CIFAR-10 are quite different from CUB-200.  

The default CIFAR-10 training hyperparameters are:
```
--model wideresnetleaky --depth 28 --widen_factor 2  --lr 0.001 --weight_decay 0.0005 --epochs 200 --l_batch_size 64 --ul_batch_size 448 
```
In some of the algorithms, you need to set `--bn_momentum 0.1`  or  `--bn_momentum 0.001`.

The default FixMatch hyperparameters are:
```
--coef 1.0 --alg pl --strong_aug true --threshold 0.95 --ema_teacher true  --ema_teacher_factor 0.999 --cutout_size 0.5
```

As an example, please try the following settings for pre-trained FixMatch on CIFAR-10:
```
pretrain_path="ckpt"
data_root="data"
dataset=cifar10
num_labels=40
arc=0 
akc=0 
CUDA_VISIBLE_DEVICES=0 nohup  python -u main.py \
--data_root $data_root --dataset $dataset --num_labels $num_labels --pretrained_weight_path $pretrain_path  \
--model wideresnetleaky --depth 28 --widen_factor 2 --l_batch_size 64 --ul_batch_size 448 \   
--lambda_mmd $arc --lambda_kd $akc  --lr 0.001 --weight_decay 0.0005 --epochs 200  \
--coef 1.0 --alg pl --strong_aug true --threshold 0.95 --ema_teacher true  --ema_teacher_factor 0.999 \
--bn_momentum 0.1  --interleave 0  --seed 10 \
>pretrained_fixmatch_${dataset}_${num_labels}.nohups 2>&1 &
```

And it is helpful to fine-tune related hyperparameters for getting better results.

**Comment by YUE-FAN** [comment]
The reproduction was successful with your hyper-parameters. Thanks a lot 👍  `--bn=0.1` gives 89.32% and `--bn=0.001` gives 87.32% in my case.

I was a bit surprised that the moment value in BN layers can be so influential on the transfer performance. Just out of curiosity, do you have any ideas or intuition about this :D

**Comment by Walleclipse** [comment]
Hi,

After tuning `lr`,  `weight_decay`, and `epochs`, you can get higher than 90% accuracy on CIFAR-10. Of course, the random choice of the 40 labeled samples could severely impact the results.

I am not very clear about the reason. But momentum in the BN layer is very critical in semi-supervised learning. For training from scratch, the small momentum with a larger epoch could achieve better performance (Just like the original [FixMatch](https://github.com/google-research/fixmatch) paper). While for training from the pretraining checkpoint, slightly bigger momentum with a slightly smaller epoch number may be better.  

About the BN layer, you can check some issue in the FixMatch repo [here](https://github.com/google-research/fixmatch/issues/20) and [here](https://github.com/google-research/fixmatch/issues/10)

**Comment by YUE-FAN** [comment]
Hi,

You are right. Running `--bn=0.1` with another seed gives 91.15% test accuracy.

That's very interesting and good to know :) Maybe it's due to the domain shift between the upstream and the downstream task (although CIFAR-10 is a subset of ImageNet classes), a larger momentum could help to quickly adapt the statistics from the pre-training task to the downstream task. I don't know. Maybe if we don't load the BN statistics during the loading of the pre-trained model, a small momentum can also achieve similar performance.

**Comment by Walleclipse** [comment]
Hi,
It is interesting to find the property of BN momentum. It is also meaningful to study the different BN momentum with and w/o loading from pre-trained checkpoints. But please notify that small BN momentum requires more training epochs.

**Comment by YUE-FAN** [comment]
Hi,

Right. Thanks for the suggestion and the discussions. I'm closing the issue now before I get any interesting results :)

**Comment by JoyHuYY1412** [comment]
> Hi,
> 
> Thanks for the great work. I have a quick question regarding the results in table 7.
> 
> Directly transferring your training protocol from CUB 200, I managed to reproduce your MixMatch results on CIFAR-10. But my attempt to reproduce the 9.05% test accuracy on CIFAR-10 with 40 labels using FixMatch was not successful. I think I must have missed some important hyper-parameters. Would you mind provide more details on that experiment?

@YUE-FAN , Hi may I ask the setting you used for reproducing MixMatch on Cifar10? 
Is the same as CUB:
`--coef 500 --alpha 0.75 --alg ict --consistency ms --warmup_iter 4000 --ema_teacher true --ema_teacher_train true --ema_teacher_warmup true --ema_teacher_factor 0.999` ?

**Comment by Walleclipse** [comment]
Hi, 
For MixMatch parameters, you can try the same parameters as CUB. But please note that the training parameters of CIFAR-10 are quite different from CUB:
`--model wideresnetleaky --depth 28 --widen_factor 2  --lr 0.001 --weight_decay 0.0005 --epochs 200 --l_batch_size 64 --ul_batch_size 448` 

If the results is not very good, you can slightly modify the '--coef' (100 or 500) and '--alpha' (0.75 or 0.5), i.e.: 
`--coef 100 --alpha 0.75 --alg ict --consistency ms --warmup_iter 4000 --ema_teacher true --ema_teacher_train true --ema_teacher_warmup true --ema_teacher_factor 0.999`

**Comment by YUE-FAN** [comment]
Hi,

It's different from the one used in CUB. I actually used a hybrid implementation myself for MixMatch on CIFAR-10.

Here is the set of hyper-parameters I used:
`--n-labeled 40 --optimizer Adam --batch-size 128 --rampupepochs 16 --lr 0.00008 --wd 0 --epochs 200 --ema_decay 0.999 --train-iteration 1024 --alpha 0.75 --lambda-u 75 --T 0.5`

Batch size is the same for the labeled and the unlabeled dataloaders. train-iteration means there are 1024 optimization steps within each epoch. Other hyper-parameters such as rampupepochs or lambda-u follow https://github.com/google-research/mixmatch 

I got 17.23% error rate after 200 epochs (it is slightly better, I think, due to a different choice of 40 labels).
